crates/ruma-common/src/canonical_json.rs

```diff
@@ -239,18 +237,48 @@ pub fn redact_content_in_place(
     object: &mut CanonicalJsonObject,
     version: &RoomVersionId,
     event_type: impl AsRef<str>,
-) {
-    object_retain_keys(object, allowed_content_keys_for(event_type.as_ref(), version));
+) -> Result<(), RedactionError> {
+    object_retain_keys(object, allowed_content_keys_for(event_type.as_ref(), version))
 }
 
-fn object_retain_keys(object: &mut CanonicalJsonObject, keys: &[&str]) {
-    let mut old_content = mem::take(object);
+fn object_retain_keys(
+    object: &mut CanonicalJsonObject,
+    allowed_keys: &AllowedKeys,
+) -> Result<(), RedactionError> {
```

Comment:
The new function has a lot of nesting, can you preserve the previous function as `object_retain_some_keys` and call it from `object_retain_keys`?